Render display names rather than usernames in reply email notifications

robertknight opened the pull request:
Render display names rather than usernames in reply email notifications if the user has set one. This is important particularly for third party accounts which have generated usernames not designed to be
human-friendly.

Slack discussion: https://hypothes-is.slack.com/archives/C2N6HPG2G/p1512663476000485

seanh commented on tests/h/models/user_test.py:
```diff
@@ -174,6 +174,19 @@ def test_userid_in_query_with_only_invalid_userid(db_session):
+    fred = models.User(authority='hypothes.is',
+                       username='fred',
+                       display_name='Fred Perry')
+    assert fred.rendered_name == 'Fred Perry'
```
I think if you use `factories` here to create the `User` then you won't have to specify `authority` (which isn't relevant to this test)

fatbusinessman commented on h/templates/emails/reply_notification.html.jinja2:
```diff
@@ -1,5 +1,5 @@
 <p>
-  <a href="{{ reply_user_url }}">{{ reply_user.username }}</a>
+  <a href="{{ reply_user_url }}">{{ reply_user.rendered_name }}</a>
```
All the templates still seem to be using `rendered_name`, despite it having changed elsewhere. Given all the tests are still passing, does this mean that this bit of functionality isn’t covered by the tests?

robertknight replied:
Gah, sorry. That is really sloppy of me. Indeed we don't have automated tests for the output of these templates. Like many views, there are just tests that the expected context is generated for the template.

robertknight replied:
Ah wait. We _do_ have tests that render these templates (`test_jinja_templates_render`), but it looks like they don't catch this class of error.

fatbusinessman commented on tests/h/emails/reply_notification_test.py:
```diff
@@ -109,10 +111,10 @@ def test_returns_text_and_body_results_from_renderers(self,
         _, subject, _, _ = generate(pyramid_request, notification)
 
-        assert subject == 'ron has replied to your annotation'
+        assert subject == '{} has replied to your annotation'.format(reply_user.display_or_user_name)
```
This test would still pass if the code used `display_name`, because in our current test fixture the `display_name` and `display_or_user_name` return equivalent values. This feels like there’s something missing, be that an extra test case or a smaller interface.

seanh commented:
I think there's at least one other place in the code (the sidebar on user pages) that uses the display name or, if there isn't one, the username. The logic is probably duplicated there.

robertknight commented:
> Given the mail templates are currently only using their parent_user and reply_user context variables to look up a name for those users, I think we can sidestep this problem by working out those names in the generate function and passing them directly into the template.

Agreed. I've simplified this PR to just determine the display names locally and removed the convenience property from the `User` model.

We haven't come to a clear agreement internally if we're going to replace usernames with display names widely, or whether it will be somewhat more context specific. So for the moment, I've just localised this change to the reply notification emails.

fatbusinessman commented:
For future look-up-ability: this is a fix for https://github.com/hypothesis/h/issues/4721.
